pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,288 @@
+	for {
+		select {
+		case <-ctx.Done():
+			t.wgTelemetryRoutine.Done()
+			return
+		case <-time.After(time.Minute * 55):
+		}
+		t.metricMux.RLock()
+		if len(t.metrics) == 0 {
+			t.metricMux.RUnlock()
+			continue
+		}
+		newStats := telemetryStats{}
+		if err := t.metricsVisitor(&newStats); err != nil {
+			log.Debug("msg", "error extracting underlying value of a stats metric", "error", err.Error())
+			continue
+		}
+		if err := syncInfoTable(t.conn, t.uuid, newStats); err != nil {
+			log.Debug("msg", "syncing new stats", "error", err.Error())
+		}
+	}
+}
+
+// metricsVisitor fills the provided newStats with most recent metric values.
+func (t *engine) metricsVisitor(newStats *telemetryStats) error {
+	t.metricMux.RLock()
+	defer t.metricMux.RUnlock()
+	if len(t.metrics) == 0 {
+		return nil
+	}
+	for columnName, metric := range t.metrics {
+		underlyingValue, err := extractMetricValue(metric)
+		if err != nil {
+			return fmt.Errorf("extracting metric value of stat '%s': %w", columnName, err)
+		}
+		switch columnName {
+		case "telemetry_metrics_ingested_samples":
+			newStats.samplesIngested = underlyingValue
+		case "telemetry_metrics_queries_failed":
+			newStats.promqlQueriesFailed = underlyingValue
+		case "telemetry_metrics_queries_executed":
+			newStats.promqlQueriesExecuted = underlyingValue
+		case "telemetry_metrics_queries_timed_out":
+			newStats.promqlQueriesTimedout = underlyingValue
+		case "telemetry_traces_queries_executed":
+			newStats.traceQueriesExecuted = underlyingValue
+		case "telemetry_traces_dependency_queries_executed":
+			newStats.traceDependencyQueriesExecuted = underlyingValue
+		default:
+			log.Error("msg", fmt.Sprintf("unregistered column name: %s", columnName))
+		}
+	}
+	return nil
+}
+
+func extractMetricValue(metric prometheus.Metric) (float64, error) {
+	var internal io_prometheus_client.Metric
+	if err := metric.Write(&internal); err != nil {
+		return 0, fmt.Errorf("error writing metric: %w", err)
+	}
+	if internal.Gauge != nil {
+		return internal.Gauge.GetValue(), nil
+	} else if internal.Counter != nil {
+		return internal.Counter.GetValue(), nil
+	}
+	return 0, fmt.Errorf("both Gauge and Counter are nil")
+}
+
+// RegisterMetric registers a counter or gauge metric for telemetry purpose.
+func (t *engine) RegisterMetric(columnName string, metric prometheus.Metric) error {
+	t.metricMux.RLock()
+	_, registered := t.metrics[column(columnName)]
+	if registered {
+		t.metricMux.RUnlock()
+		return fmt.Errorf("metric already registered for column '%s'", columnName)
+	}
+	t.metricMux.RUnlock()
+
+	ok, err := isCounterOrGauge(metric)
+	if err != nil {
+		return fmt.Errorf("is counter or gauge: %w", err)
+	}
+	if !ok {
+		return fmt.Errorf("metric should only be of type Counter or Gauge")
+	}
+
+	t.metricMux.Lock()
+	defer t.metricMux.Unlock()
+	t.metrics[column(columnName)] = metric
+	return nil
+}
+
+func isCounterOrGauge(metric prometheus.Metric) (bool, error) {
```

Comment:
Instead of writing to a Metric, why not check underlying type is `prometheus.Gauge` or `prometheus.Counter`?